### PR TITLE
Geode-7600-3

### PR DESCRIPTION
**Pull request by mhansonp**

### CODE REVIEW

**Comment by mhansonp** [comment]
Submitting a bug against DistributedAckRegionCCEDUnitTest > testClearOnNonReplicateWithConcurrentEvents this is a product issue.

File: geode-core/src/distributedTest/java/org/apache/geode/cache/ConnectionPoolDUnitTest.java [hunk 1]
```diff
@@ -2720,8 +2677,10 @@ public void run2() throws CacheException {
     });
 
     vm2.invoke("Verify destroy propagate", () -> {
-      Region<Object, Object> region = getRootRegion().getSubregion(name);
-      await().until(() -> region == null);
+      await().until(() -> {
+        logger.info("MLH region = " + getRootRegion().getSubregion(name));
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache/ConnectionPoolDUnitTest.java]: Remember to remove these `logger` statements containing `MLH`

File: geode-core/src/distributedTest/java/org/apache/geode/cache/ConnectionPoolDUnitTest.java [hunk 2]
```diff
@@ -3207,7 +3149,7 @@ public void test028DynamicRegionCreation() {
       assertThat(r).isNotNull();
       String drName = r.getFullPath() + Region.SEPARATOR + dynFromServerName;
 
-      await().pollInterval(100, MILLISECONDS).until(() -> getCache().getRegion(drName) == null);
+      await().until(() -> getCache().getRegion(drName) == null);
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache/ConnectionPoolDUnitTest.java]: I think this test will run into same thing I hit with another test. There's a brief window in which `getRegion` may throw `RegionDestroyedException` before it starts returning null:
```
await().ignoreException(RegionDestroyedException.class).until(() -> 
    getCache().getRegion(drName) == null);
```

File: geode-core/src/distributedTest/java/org/apache/geode/cache/ConnectionPoolDUnitTest.java [hunk 3]
```diff
@@ -3217,13 +3159,11 @@ public void test028DynamicRegionCreation() {
       Region<Object, Object> r = getRootRegion(name);
       assertThat(r).isNotNull();
       String drName = r.getFullPath() + Region.SEPARATOR + dynFromServerName;
-      await().pollInterval(100, MILLISECONDS).until(() -> getCache().getRegion(drName) == null);
+      await().until(() -> getCache().getRegion(drName) == null);
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache/ConnectionPoolDUnitTest.java]: Add `ignoreException(RegionDestroyedException.class)`

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java [hunk 4]
```diff
@@ -147,129 +146,118 @@ public void testConcurrentOpWithGII() {
...
+            @Override
+            public void vMotionDuringGII(Set recipientSet, LocalRegion region) {
+              InitialImageOperation.VMOTION_DURING_GII = false;
+              final InitializationLevel oldLevel =
+                  LocalRegion.setThreadInitLevelRequirement(BEFORE_INITIAL_IMAGE);
+              LocalRegion ccregion = (LocalRegion) cache.getInternalRegionByPath("/" + name);
+              try {
+                // wait for the update op (sent below from vm1) to arrive, then allow the GII to
+                // happen
+                while (!ccregion.isDestroyed() && ccregion.getRegionEntry(key) == null) {
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java]: I recommend changing any sort of loop containing a sleep into an `await` call:
```
await().ignoreException(RegionDestroyedException.class).until(() ->
    ccregion.isDestroyed() && ccregion.getRegionEntry(key) != null);
```

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java [hunk 5]
```diff
@@ -147,129 +146,118 @@ public void testConcurrentOpWithGII() {
...
+      // create a fake member ID that will be < mine and lose a concurrency check
+      MemberData nm =
+          CCRegion.getDistributionManager().getDistributionManagerId().getMemberData();
+      InternalDistributedMember mbr = null;
+      try {
+        mbr = new InternalDistributedMember(nm.getInetAddress().getCanonicalHostName(),
+            nm.getMembershipPort() - 1, "fake_id", "fake_id_ustring",
+            ClusterDistributionManager.NORMAL_DM_TYPE, null, null);
+        tag.setMemberID(mbr);
+      } catch (UnknownHostException e) {
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java]: Delete any `catch` block with a `fail` and let the test throw `UnknownHostException`.

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java [hunk 6]
```diff
@@ -312,63 +299,47 @@ public void testTombstoneExpirationRace() {
...
+      // reaped by distributed-GC
+      RegionEntry entry = CCRegion.getRegionEntry("cckey0");
+      VersionTag tag = entry.getVersionStamp().asVersionTag();
+      assertThat(tag.getEntryVersion()).isGreaterThan(1);
+      tag.setVersionTimeStamp(
+          System.currentTimeMillis() - TombstoneService.REPLICATE_TOMBSTONE_TIMEOUT - 1000);
+      entry.getVersionStamp().setVersionTimeStamp(tag.getVersionTimeStamp());
+      try {
+        entry.makeTombstone(CCRegion, tag);
+      } catch (RegionClearedException e) {
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java]: Delete this `catch` block and change the methods to `throws Exception`.

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionCCEDUnitTest.java [hunk 7]
```diff
@@ -163,78 +150,58 @@ public void testTombstones() {
...
       public void run() {
         try {
           RegionFactory f = getCache().createRegionFactory(getRegionAttributes());
           CCRegion = (LocalRegion) f.create(name);
+
+          // noinspection OverwrittenKey
           CCRegion.put("cckey0", "ccvalue");
+          // noinspection OverwrittenKey
           CCRegion.put("cckey0", "ccvalue"); // version number will end up at 4
         } catch (CacheException ex) {
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionCCEDUnitTest.java]: Optional: Delete this `catch` block. `CacheException` is a `RuntimeException` so it doesn't even requires a `throws` clause.

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionCCEDUnitTest.java [hunk 8]
```diff
@@ -163,78 +150,58 @@ public void testTombstones() {
...
+      // means it could be in the process of being reaped by distributed-GC
+      RegionEntry entry = CCRegion.getRegionEntry("cckey0");
+      VersionTag tag = entry.getVersionStamp().asVersionTag();
+      assertTrue(tag.getEntryVersion() > 1);
+      tag.setVersionTimeStamp(
+          System.currentTimeMillis() - TombstoneService.REPLICATE_TOMBSTONE_TIMEOUT - 1000);
+      entry.getVersionStamp().setVersionTimeStamp(tag.getVersionTimeStamp());
+      try {
+        entry.makeTombstone(CCRegion, tag);
+      } catch (RegionClearedException e) {
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionCCEDUnitTest.java]: Another `catch` block to delete.

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionCCEDUnitTest.java [hunk 9]
```diff
@@ -163,78 +150,58 @@ public void testTombstones() {
...
     // perform a CREATE with a new version number
-    vm0.invoke(localDestroyRunnable);
-    vm1.invoke(new SerializableRunnable("Check that the create() was applied") {
-      @Override
-      public void run() {
-        RegionEntry entry = CCRegion.getRegionEntry("cckey0");
-        assertThat(entry.getVersionStamp().getEntryVersion()).isEqualTo(1);
+    vm0.invoke("Locally destroy the entry and do a create that will be propagated with v1", () -> {
+      CCRegion.getRegionMap().removeEntry("cckey0", CCRegion.getRegionEntry("cckey0"), true);
+      if (CCRegion.getRegionEntry("ckey0") != null) {
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionCCEDUnitTest.java]: Replace this if and fail block with an assertion:
```
assertThat(CCRegion.getRegionEntry("ckey0")).isNull();
```

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java [hunk 10]
```diff
@@ -71,71 +71,58 @@
...
-          } catch (IllegalStateException ex) {
-            // pass...
-            assertThat(getRootRegion()).isNull();
-          }
-
-        } catch (CacheException ex) {
-          fail("While creating GLOBAL Region", ex);
+          Region<Object, Object> rootRegion = factory.create("INCOMPATIBLE_ROOT");
+          fail("Should have thrown an IllegalStateException");
+          factory.createSubregion(rootRegion, name);
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java]: I don't think the code can ever reach lines 116 or 117

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java [hunk 11]
```diff
@@ -240,80 +206,66 @@ public void uncaughtException(Thread t, Throwable e) {
         };
 
         Thread[] threads = new Thread[threadsPerVM];
-        for (int i = 0; i < threadsPerVM; i++) {
-          Thread thread = new Thread(group, new Runnable() {
-            @Override
-            public void run() {
+        for (int i1 = 0; i1 < threadsPerVM; i1++) {
+          Thread thread = new Thread(group, () -> {
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java]: Optional: We should try to replace all testing `Threads` like these with use of `ExecutorServiceRule`. Spinning up our own `Thread` should be considered `deprecated` and replaced with `Executors`. The `ExecutorServiceRule` will report a failure if any thread is left running after the test or if any of these threads threw anything.

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java [hunk 12]
```diff
@@ -240,80 +206,66 @@ public void uncaughtException(Thread t, Throwable e) {
...
-                throw e;
-              } catch (Throwable t) {
-                logger
-                    .info("testSynchronousIncrements." + this + " caught Throwable", t);
+
+              } catch (IllegalStateException | InterruptedException | CacheLoaderException
+                  | CacheWriterException | TimeoutException ex) {
+                fail("While incrementing", ex);
               }
+            } catch (VirtualMachineError e) {
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java]: Optional: Delete all catch blocks that catch Error or Throwable.

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java [hunk 13]
```diff
@@ -240,80 +206,66 @@ public void uncaughtException(Thread t, Throwable e) {
...
+
+                  lock.unlock();
                 }
-              } catch (VirtualMachineError e) {
-                throw e;
-              } catch (Throwable t) {
-                logger
-                    .info("testSynchronousIncrements." + this + " caught Throwable", t);
+
+              } catch (IllegalStateException | InterruptedException | CacheLoaderException
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java]: Optional: Remove catch block containing fail and add throws clause to method(s).

File: geode-core/src/distributedTest/java/org/apache/geode/cache/ConnectionPoolDUnitTest.java (hunk 2 above)
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/cache/ConnectionPoolDUnitTest.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/cache/ConnectionPoolDUnitTest.java (hunk 3 above)
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/cache/ConnectionPoolDUnitTest.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java (hunk 4 above)
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java (hunk 5 above)
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java (hunk 6 above)
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/cache30/DistributedAckRegionCCEDUnitTest.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionCCEDUnitTest.java (hunk 7 above)
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionCCEDUnitTest.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionCCEDUnitTest.java (hunk 8 above)
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionCCEDUnitTest.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionCCEDUnitTest.java (hunk 9 above)
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionCCEDUnitTest.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java (hunk 10 above)
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java]: true. Deleted.

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java (hunk 12 above)
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java (hunk 13 above)
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/cache30/GlobalRegionDUnitTest.java]: done

**Pull request merged**